### ISSUE TITLE
How to get the top 20 assets with the highest increase within 24 hours

### ISSUE THREAD

**Issue by WinPoss** [freqtrade/freqtrade#10348]
## Describe your environment

* Operating system: ubuntu 20.04.6
* Python Version: 3.12
* CCXT version: 4.3.11
* Freqtrade Version: 2024.4
  
## Your question

*Ask the question you have not been able to find an answer in the [Documentation](https://www.freqtrade.io/en/latest/)*

My current configuration file is set up like this:

```
"pairlists": [
        {
            "method": "VolumePairList",
            "number_assets": 50,
            "sort_key": "quoteVolume",
            "min_value": 0,
            "refresh_period": 1800
        }
    ],
```
What you get is the TOP 50 assets ranked according to trading volume.

Is there a setting method that allows me to live with the top 20 assets that have increased in the last 24 hours? 

Thanks.

**Comment by stash86** [comment]
There is no such pairlist plugin for that yet. Someone might create it in the future. Gonna keep this open for now

**Comment by winposs** [comment]
Okay, this is a practical function, looking forward to your updates.

**Comment by WinPoss** [comment]
> There is no such pairlist plugin for that yet. Someone might create it in the future. Gonna keep this open for now

Okay, this is a practical function, looking forward to your updates.

**Comment by WinPoss** [comment]
It’s a setting similar to this:

```
"pairlists": [
        {
            "method": "ChangePairList",
            "number_assets": 20,
            "sort_key": "change",
            "min_change": 0,
            "refresh_period": 1800
        }
    ],
```

**Comment by xmatthias** [comment]
As the label already mentions - it's a good issue to get started contributing to freqtrade.

This also (implicitly) means that it's unlikely to be picked up by experienced contributors (unless they're personally interested in it, obviously) - as it's good practice to leave simple issues for new contributors as good "entry point".

---

The best way to get started will be to start by copying an existing, similar pairlist, and then implement the logic you desire there.
We've also got a documentation for contributing [pairlists](https://www.freqtrade.io/en/latest/developer/#pairlists) - which you might want to read through to understand what the individual functions should be doing.

We look forward to a Pull request on this topic :+1:

**Comment by WinPoss** [comment]
> As the label already mentions - it's a good issue to get started contributing to freqtrade.
> 
> This also (implicitly) means that it's unlikely to be picked up by experienced contributors (unless they're personally interested in it, obviously) - as it's good practice to leave simple issues for new contributors as good "entry point".
> 
> The best way to get started will be to start by copying an existing, similar pairlist, and then implement the logic you desire there. We've also got a documentation for contributing [pairlists](https://www.freqtrade.io/en/latest/developer/#pairlists) - which you might want to read through to understand what the individual functions should be doing.
> 
> We look forward to a Pull request on this topic 👍

I don't know programming and I don't speak English. Usually, I ask ChatGPT to help me write my strategy files for freqtrade. I also use ChatGPT to translate when I ask for help on GitHub. I feel that developing a PairList is too difficult for me. I hope there are experts in the community who need this functionality and can develop it. 

Wishing everyone a happy life.

**Comment by WinPoss** [comment]
[ChangePairList.zip](https://github.com/user-attachments/files/15934818/ChangePairList.zip)

I have written a `ChangePairList.py` file, but I have three questions that I would like to ask:

1. Is my file correct? Can it provide a list of trading pairs based on the price changes of assets?

2. If the answer to question 1 is yes, and my strategy is set with a trading period of 5 minutes, can I obtain the asset list based on the price fluctuations of the assets over a 24-hour period instead of the 5-minute period?

3. It seems that I did not see the functionality to request a list of trading pairs from the exchange in `ChangePairList.py`. How can I make this file work?

Thank you.

**Comment by froggleston** [comment]
It's better that you submit a PR - this way things can be iteratively developed and tested. Downloading random zip files is potentially dangerous so it would be good to try to avoid it. 

You can also paste any code directly in these issues using the triple backticks.

**Comment by xmatthias** [comment]
A guide for the github PR workflow can be found [here](https://guides.github.com/introduction/flow/).

Youll find that its not difficult to create a Pull request, but will allow "per line" comments, avoiding random "ah that comment means that particular line" headscratching

**Comment by xmatthias** [comment]
@winposs i don't think your file will work (aside from the fact that it should be a pull request, not a zip file ...).

You'll be comparing absolute values - which won't make sense and will always have "higher-priced" assets at the top.

Looking at the ticker structure - `"change"` is the absolute change ... for BTC today, that's ~3000 (5.8%)
for ETH, it's 206.13 (6%)
for XRP - it's -0.0006 (-0.1%).
Based on your logic - BTC will be top - but ETH had the biggest relative gain.

what you'll want to look at instead is the `"percentage"` column ... though i don't think it's well defined what this change (or percentage) refers to (is it today? 24h rolling? ...). It'll surely depend per exchange.

your calculation (from candles) essentially has the same flaw.

**Comment by jainanuj94** [comment]
Hey @winposs, i took the liberty of taking your code and applying certain modifications to support `percentage` instead of `total change`. Hope that works. 

Here is a PR for the same: https://github.com/freqtrade/freqtrade/pull/10454
Please do have a look at it @winposs . 

Cheers!

**Comment by RkjConceptx** [comment]
i am also not a good programmer, but bellow code working for me
```python
import logging
from datetime import datetime, timedelta
from typing import Dict, List

from cachetools import TTLCache

from freqtrade.exceptions import OperationalException
from freqtrade.exchange import timeframe_to_prev_date
from freqtrade.exchange.types import Tickers
from freqtrade.plugins.pairlist.IPairList import IPairList, PairlistParameter
from freqtrade.util import dt_now


logger = logging.getLogger(__name__)


class CustomPriceChangePairList(IPairList):
    is_pairlist_generator = True

    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

        # Custom configuration parameters
        self._min_price_change = self._pairlistconfig.get("percent_change",0.1)

        # Ensure necessary parameters are present in pairlistconfig
        if "number_assets" not in self._pairlistconfig:
            raise OperationalException(
                "`number_assets` not specified. Please check your configuration "
                'for "pairlist.config.number_assets"'
            )

        self._number_pairs = self._pairlistconfig["number_assets"]
        self._refresh_period = self._pairlistconfig.get("refresh_period", 1800)

        # Initialize cache
        self._pair_cache = TTLCache(maxsize=1, ttl=self._refresh_period)

    @property
    def needstickers(self) -> bool:
        """
        Boolean property defining if tickers are necessary.
        If no Pairlist requires tickers, an empty Dict is passed
        as tickers argument to filter_pairlist
        """
        return True  # Since we need tickers for filtering based on price change and listing time

    def short_desc(self) -> str:
        """
        Short whitelist method description - used for startup-messages
        """
        return f"CustomPriceChangePairList - top {self._number_pairs} pairs with 10% price change and listed for 3 days."

    @staticmethod
    def description() -> str:
        return "Provides custom pair list based on 10% price change and listing time."

    @staticmethod
    def available_parameters() -> Dict[str, PairlistParameter]:
        return {
            "number_assets": {
                "type": "number",
                "default": 30,
                "description": "Number of assets",
                "help": "Number of assets to use from the pairlist",
            },
            **IPairList.refresh_period_parameter(),
        }

    def gen_pairlist(self, tickers: Tickers) -> List[str]:
        """
        Generate the initial pairlist
        :param tickers: Tickers (from exchange.get_tickers). May be cached.
        :return: List of pairs
        """
        pairlist = self._pair_cache.get("pairlist")
        if pairlist:
            return pairlist.copy()
        else:
            pairlist = list(tickers.keys()) if tickers else []
            self._pair_cache["pairlist"] = pairlist.copy()
            return pairlist

    def filter_pairlist(self, pairlist: List[str], tickers: Tickers) -> List[str]:
        """
        Filters the pairlist based on custom criteria (10% price change and 3 days listing).
        :param pairlist: List of pairs to filter
        :param tickers: Tickers (from exchange.get_tickers). May be cached.
        :return: Filtered list of pairs
        """
        filtered_pairs = []

        for pair in pairlist:
            if self._check_price_change(pair, tickers):
                filtered_pairs.append(pair)

        return filtered_pairs

    def _check_price_change(self, pair: str, tickers: Tickers) -> bool:
        """
        Checks if the pair has had a minimum 10% price change.
        :param pair: Pair symbol
        :param tickers: Tickers (from exchange.get_tickers). May be cached.
        :return: True if the pair meets the price change criteria, False otherwise
        """
        if pair not in tickers:
            return False

        ticker = tickers[pair]
        if "open" not in ticker or "close" not in ticker:
            return False

        open_price = ticker["open"]
        close_price = ticker["close"]

        # Calculate price change percentage
        price_change = (close_price - open_price) / open_price

        return price_change >= self._min_price_change

```

**Comment by xmatthias** [comment]
A guide for the github PR workflow can be found [here](https://guides.github.com/introduction/flow/).

An explanation as to why you should use feature branches can be found [here](https://hynek.me/articles/pull-requests-branch/).

---

That said - there's already a Pull request available for this in #10454 ... so i don't think we'll need another PR adding the same thing.
